File: app/src/main/java/org/mifos/mobile/ui/adapters/RecentTransactionListAdapter.java
package org.mifos.mobile.ui.adapters;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import org.mifos.mobile.R;
import org.mifos.mobile.injection.ActivityContext;
import org.mifos.mobile.models.Transaction;
import org.mifos.mobile.utils.CurrencyUtil;
import org.mifos.mobile.utils.DateHelper;
import org.mifos.mobile.utils.Utils;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import androidx.recyclerview.widget.RecyclerView;
import butterknife.BindView;
import butterknife.ButterKnife;

/**
 * @author Vishwajeet
 * @since 10/08/16
 */
public class RecentTransactionListAdapter extends
        RecyclerView.Adapter<RecentTransactionListAdapter.ViewHolder> {

    private List<Transaction> transactions;
    private Context context;

    @Inject
    public RecentTransactionListAdapter(@ActivityContext Context context) {
        transactions = new ArrayList<>();
        this.context = context;
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.row_recent_transaction, parent, false);
        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(ViewHolder holder, int position) {
        Transaction transaction = getItem(position);
        String currencyRepresentation = transaction.getCurrency().getDisplaySymbol();
        if (currencyRepresentation == null) {
            currencyRepresentation = transaction.getCurrency().getCode();
        }
        holder.tvAmount.setText(context.getString(R.string.string_and_string,
                currencyRepresentation, CurrencyUtil.formatCurrency(context,
                        transaction.getAmount())));
        holder.tvTypeValue.setText(Utils.formatTransactionType(transaction.getType().getValue()));
        holder.tvTransactionsDate.setText(DateHelper.getDateAsString(transaction.
                getSubmittedOnDate()));
    }

    public void setTransactions(List<Transaction> transactions) {
        this.transactions = transactions;
        notifyDataSetChanged();
    }


    public Transaction getItem(int position) {
        return transactions.get(position);
    }

    @Override
    public int getItemCount() {
        return transactions.size();
    }

    public static class ViewHolder extends RecyclerView.ViewHolder {

        @BindView(R.id.tv_transactionDate)
        TextView tvTransactionsDate;

        @BindView(R.id.tv_amount)
        TextView tvAmount;

        @BindView(R.id.tv_value)
        TextView tvTypeValue;

        public ViewHolder(View v) {
            super(v);
            ButterKnife.bind(this, v);
        }
    }
}

File: app/src/main/java/org/mifos/mobile/ui/adapters/RecentTransactionListAdapter.kt
package org.mifos.mobile.ui.adapters

import android.content.Context
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import org.mifos.mobile.MifosSelfServiceApp.Companion.context
import org.mifos.mobile.R
import org.mifos.mobile.databinding.RowRecentTransactionBinding
import dagger.hilt.android.qualifiers.ActivityContext


import org.mifos.mobile.models.Transaction
import org.mifos.mobile.models.client.Type
import org.mifos.mobile.utils.CurrencyUtil.formatCurrency
import org.mifos.mobile.utils.DateHelper.getDateAsString
import org.mifos.mobile.utils.Utils.formatTransactionType
import javax.inject.Inject

/**
 * @author Vishwajeet
 * @since 10/08/16
 */
class RecentTransactionListAdapter @Inject constructor(@ActivityContext context: Context) :
    RecyclerView.Adapter<RecentTransactionListAdapter.ViewHolder>() {

    private var transactions: MutableList<Transaction?>
    private val context: Context

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val binding =
            RowRecentTransactionBinding.inflate(LayoutInflater.from(parent.context), parent, false)
        return ViewHolder(binding)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val (_, _, _, type, _, currency, amount, submittedOnDate) = getItem(position)
        var currencyRepresentation = currency?.displaySymbol ?: currency?.code
        if (currencyRepresentation == null) {
            currencyRepresentation = currency?.code
        }

        holder.bind(currencyRepresentation, type, submittedOnDate, amount)
    }

    fun setTransactions(transactions: MutableList<Transaction?>?) {
        if (transactions != null) {
            this.transactions = transactions
        }
        notifyDataSetChanged()
    }

    fun getItem(position: Int): Transaction {
        return transactions[position]!!
    }

    override fun getItemCount(): Int {
        return transactions.size
    }

    class ViewHolder(private val binding: RowRecentTransactionBinding) :
        RecyclerView.ViewHolder(binding.root) {
        fun bind(
            currencyRepresentation: String?,
            type: Type,
            submittedOnDate: List<Int>,
            amount: Double?,
        ) {
            with(binding) {
                tvAmount.text = context?.getString(
                    R.string.string_and_string,
                    currencyRepresentation,
                    formatCurrency(
                        context,
                        amount!!,
                    ),
                )
                tvValue.text = formatTransactionType(type.value)
                tvTransactionDate.text = getDateAsString(submittedOnDate)
            }
        }
    }

    init {
        transactions = ArrayList()
        this.context = context
    }
}
